driver-core/src/main/com/mongodb/internal/connection/AwsAuthenticator.java

```diff
@@ -223,20 +223,18 @@ private String getPassword() {
 
         @Nullable
         private String getSessionToken() {
-            String token = credential.getMechanismProperty("AWS_SESSION_TOKEN", null);
-            if (credential.getPassword() == null || credential.getUserName() == null) {
-                if (token == null) {
-                    if ((token = System.getenv("AWS_SESSION_TOKEN")) == null) {
-                        token = BsonDocument
-                                .parse(getHttpResponse())
-                                .getString("Token")
-                                .getValue();
-                    }
-                } else {
-                    throw new IllegalArgumentException("The connection string contains auth properties and no username and password");
-                }
+            if (credential.getPassword() != null && credential.getUserName() != null) {
```

Comment:
The spec says: 

> If a username is provided without a password (or vice-versa) drivers MUST raise an error.

Since this method should only be called after that check has been made, this check can be simplified to:

    credential.getUserName() != null